### ISSUE TITLE
How to Fine-tune Rostlab/prot_t5_xl_uniref50 Model for Sequence Generation

### ISSUE THREAD

**Issue by gundalav** [agemagician/ProtTrans#111]
Dear Sir,

Firstly, I want to thank you for the valuable assistance you have provided me in the past. Your expertise has been immensely helpful, and I find myself in need of your guidance once again.

I'd like to fine-tune the pre-trained model `Rostlab/prot_t5_xl_uniref50`. 

The dataset I plan to use for fine tuning looks like this (colon separated):

```
P R T <extra_id_0> I N S E Q W <extra_id_1> E N C E :  P R T K I N S E Q W H E N C E 
G M M  <extra_id_0>  <extra_id_1> K P H G : G M M  V E K P H G
R H G L  <extra_id_0>  <extra_id_1> : R H G L  Q F

...etc...
```

First column is the input from user and the output is on the right.
The output is simply the 'active' protein inferred from experiments.

Unfortunately, I have been unable to find a suitable example in the [ProtTrans repository](https://github.com/agemagician/ProtTrans/tree/master/Fine-Tuning).

Additionally, I am curious if the Hugging Face script [run_mlm.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_mlm.py) can be utilized with the aforementioned pre-trained model.

I would truly appreciate your insights and recommendations on how to proceed with this task. Thank you in advance for your time and consideration.

Sincerely,
G.V.

**Comment by mheinzinger** [comment]
Hey,

I see. You should probably follow the T5 span-corruption convention and put <extra_id> tokens in the output to give the model some guidance on what to generate. This way, you are closer to the original pre-training objective which should make learning easier. So for example, the first example shuold read: `P R T <extra_id_0> I N S E Q W <extra_id_1> E N C E :  <s> <extra_id_0> K <extra_id_0> H <\s>`
I am also not sure whether the example you linked above would work more-or-less out-of-the-box because it is for pure masked language modeling as done in e.g. BERT but not for span-corruption as done in T5. I understand that the two are essentially identical if you choose span-length of 1 but this is just a decision we made. Once you increase span-length, you see how the two differ. You can probably force T5 also to do masked-language modeling but then you also only need the T5-Encoder without the decoder. So if you remove the T5-decoder you might be able to run the example you linked. 
If you want to stay closer to the original pre-training objective, I would rather recommend the two following scripts:
 [T5-pre-training](https://github.com/huggingface/transformers/tree/main/examples/flax/language-modeling#t5-like-span-masked-language-modeling)
 [T5-fine-tuning](https://github.com/huggingface/transformers/blob/main/examples/pytorch/summarization/run_summarization_no_trainer.py)
 Probably you would only need the very first one (pre-training) and adjust the way data is loaded (you want it to load pre-defined spans instead of randomly corrupting input)

**Comment by gundalav** [comment]
> 

@mheinzinger 


Dear Michael,

Thank you for your prompt and insightful response. Your suggestion to use [T5 pre-training](https://github.com/huggingface/transformers/tree/main/examples/flax/language-modeling#t5-like-span-masked-language-modeling) seems particularly promising, and I am eager to explore its potential.

Before I proceed, I would appreciate some further guidance on a few points:

1. For your Rostlab/prot_t5_xl_uniref50 model, what is the appropriate vocab_size to select?

2. In terms of fine-tuning, what would be a reasonable size for the dataset I should provide? Below is a sample of the format I will be working with:

```
P R T <extra_id_0> I N S E Q W <extra_id_1> E N C E :  <s> <extra_id_0> K <extra_id_0> H <\s>
...how many rows ideally?...
```

3. In the end, my code will include the following segment:

```
from transformers import T5Config

config = T5Config.from_pretrained("Rostlab/prot_t5_xl_uniref50", vocab_size=tokenizer.get_vocab_size())
config.save_pretrained("./gundala-t5-active_protein.model")

```
How does my model interact with Rostlab/prot_t5_xl_uniref50 in this instance? Specifically, does it re-weight your model?

Once again, I am grateful for your assistance and expertise. 

Sincerely,
G.V

**Comment by mheinzinger** [comment]
1. Vocab-size should be 128 if I'm not mistaken: https://huggingface.co/Rostlab/prot_t5_xl_uniref50/blob/main/config.json#L24
You can always check this yourself by doing sth like len(tokenizer) after you've loaded a pre-trained tokenizer in huggingface

2. As many rows as possible, ideally ;) after all you want to fine-tune ~3B parameters. I made good experience with a few million but I did not test extensively what happens if you have much less.

3. you should not only save the config but also the model weights after/during fine-tuning. But the pre-training script that I had sent you already handles this. there should be a line saying sth like "model.save()" at one point. This writes the current checkpoint (model weights etc) to disk. Yes, it does "re-weight" ProtT5 in a sense that it changes its weights. 

4. Currently, the pre-training script that I have linked above assumes that you input a "P R T E I N S E Q W E N C E" which then gets randomly corrupted during training ("P R <extra_id1> ..."). However, if I understood you correctly, you have specific sites that you would like to corrupt/reconstruct. If that were correct, you would have to adjust the way data is loaded and de-activate the random masking. However, maybe I also misunderstood you and you just want to fine-tune ProtT5 on a certain family. Then you are good2go with the script I have linked above. You just need to check that your input CSV is compatible with the script.

**Comment by gundalav** [comment]
@mheinzinger 

Hi Michael. 

I have a final question for today. 

As per your recommendation, I have been using the `run_t5_mlm_flax.py` script, available at this link: [T5-like Span Masked Language Modeling](https://github.com/huggingface/transformers/tree/main/examples/flax/language-modeling#t5-like-span-masked-language-modeling). I am currently working with their Norwegian-T5-Base example.

Although the script is executing as expected, I noticed that it does not seem to utilize the GPU. Is this the intended behavior, or is there a way to enable GPU usage for better performance?

Sincerely,
G.V.

**Comment by mheinzinger** [comment]
Just in case it should still be relevant, we now have a finetuning script for ProtT5 using LoRA layers: https://github.com/agemagician/ProtTrans/tree/master/Fine-Tuning

**Comment by mheinzinger** [comment]
Update: https://github.com/agemagician/ProtTrans/issues/137#issuecomment-1817576165
